mono/metadata/threads.c

```diff
@@ -4515,61 +4566,90 @@ mono_thread_execute_interruption (void)
 	mono_thread_info_clear_self_interrupt ();
 
 	/* If there's a pending exception and an AbortRequested, the pending exception takes precedence */
-	if (sys_thread->pending_exception) {
-		MonoException *exc;
-
-		exc = sys_thread->pending_exception;
-		sys_thread->pending_exception = NULL;
-
-		UNLOCK_THREAD (thread);
-		return exc;
-	} else if (thread->state & (ThreadState_AbortRequested)) {
-		UNLOCK_THREAD (thread);
-		g_assert (sys_thread->pending_exception == NULL);
-		if (thread->abort_exc == NULL) {
-			/* 
-			 * This might be racy, but it has to be called outside the lock
-			 * since it calls managed code.
-			 */
-			MONO_OBJECT_SETREF (thread, abort_exc, mono_get_exception_thread_abort ());
+	//pending_exception = sys_thread->pending_exception;
+	MONO_HANDLE_GET (pending_exception, sys_thread, pending_exception);
+	if (!MONO_HANDLE_IS_NULL (pending_exception)) {
+		MONO_HANDLE_ASSIGN (exc, pending_exception);
+		// sys_thread->pending_exception = NULL;
+		MONO_HANDLE_SETRAW (sys_thread, pending_exception, NULL);
+		goto exit;
+	} else if (MONO_HANDLE_GETVAL (thread, state) & ThreadState_AbortRequested) {
+		g_assert (MONO_HANDLE_IS_NULL (pending_exception));
+
+		// Does the thread already have an abort exception?
+		// If not, create a new one and set it on demand.
+
+		// local_abort_exc = thread->abort_exc;
+		MONO_HANDLE_GET (local_abort_exc, thread, abort_exc);
+		if (MONO_HANDLE_IS_NULL (local_abort_exc)) {
+			// Exception creation has to be outside the lock
+			// since it calls managed code. To avoid race,
+			// drop lock, create local, retry.
+			if (!MONO_HANDLE_IS_NULL (local_abort_exc)) {
+				// thread->abort_exc = local_abort_exc;
+				MONO_HANDLE_SET (thread, abort_exc, local_abort_exc);
+			} else {
+				UNLOCK_THREAD_HANDLE (thread);
+				unlock = FALSE;
+				MONO_HANDLE_ASSIGN (local_abort_exc, mono_new_exception_thread_abort ());
+				goto retry;
+			}
 		}
-		return thread->abort_exc;
-	} else if (thread->state & (ThreadState_SuspendRequested)) {
+		// Now we have an abort exception to return, because
+		// we already did, or because we just created and set it.
+		MONO_HANDLE_ASSIGN (exc, local_abort_exc);
+		goto exit;
+	} else if (MONO_HANDLE_GETVAL (thread, state) & ThreadState_SuspendRequested) {
 		/* calls UNLOCK_THREAD (thread) */
 		self_suspend_internal ();
-		return NULL;
-	} else if (thread->thread_interrupt_requested) {
-
-		thread->thread_interrupt_requested = FALSE;
-		UNLOCK_THREAD (thread);
-		
-		return(mono_get_exception_thread_interrupted ());
+		unlock = FALSE;
+		goto return_null;
+	} else if (MONO_HANDLE_GETVAL (thread, thread_interrupt_requested)) {
+		// thread->thread_interrupt_requested = FALSE
+		MONO_HANDLE_SETVAL (thread, thread_interrupt_requested, gboolean, FALSE);
+		UNLOCK_THREAD_HANDLE (thread);
+		unlock = FALSE;
+		MONO_HANDLE_ASSIGN (exc, mono_new_exception_thread_interrupted ());
+		goto exit;
 	}
 	
-	UNLOCK_THREAD (thread);
-	
-	return NULL;
+return_null:
+	if (unlock) {
```

Comment:
Why duplicating this block of code with https://github.com/mono/mono/pull/6639/files#diff-f1a9287559d3aa620b058619f47f00c9R4624?